### ISSUE TITLE
[Feature Request] Native GCC input support

### ISSUE THREAD

**Issue by MutantAura** [Ryujinx/Ryujinx#2425]
#### Overview:
- Native support of the Gamecube controller adapter (WUP-028) in a similar fashion to Dolphin.
- Addition of a "GameCube Controller" input device with others such as pro controller, joycons etc.
- Extension of https://github.com/Ryujinx/Ryujinx/issues/1978 as this issue was not fixed for official adapters.

#### Smaller Details: 
- Would optimally support WUP-028 devices out of box after a WinUSB driver installation.
- As opposed to dolphins implementation (Only allowing native controls to be mapped to native controller) be configurable in the Input config window. Still enabling GameCube controllers to be mapped to pro controllers if needed/wanted.

#### Nature of Request:  
- Addition
  - Dolphin and other emulators use LibUSB under the GPLv2.1 license to access the WUP-028 device over USB
  - Addition of GCC input option alongside Pro controller, joycons (L+R) etc. So other controller types can also be mapped to an emulated GameCube controller for use in games that support it e.g. Smash Bros.
  
### Why would this feature be useful?
- Current workarounds are Xinput wrapping through such programs as https://github.com/Struggleton/Delfinovin or x360ce however both of these result in inaccuracies in range/deadzones of controllers and still show as Pro controllers.
- Games such as Super Smash Bros. Ultimate have specific GCC inputs in areas such as training mode to reset the characters etc that are not possible through the "Xinput -> pro controller" workaround. This is a major inconvenience for using Ryujinx with a GCC.
- Most mapping tools do not map accurately enough to simulate real switch GCC support and most of the time full input is impossible to reach through such methods.

**Comment by german77** [comment]
SDL supports the GC adapter in PC mode. As for the in game GC controller that needs to be implemented

**Comment by MutantAura** [comment]
This is true of adapters like mayflash and 3rd party that support a "PC mode" however Nintendo official Wii U usb adapters are not supported at all without specialised programs to remap their input to X or d input.

A mayflash in "Wii U" mode is the only option that official adapters are capable of outputting.

**Comment by bylaws** [comment]
@MutantAura SDL has supported the official GC adapter for the wiiu since 2.0.12

**Comment by MutantAura** [comment]
If SDL supports it then the Ryu implementation as far as I can test does not.
Below is a vid of the adapter plugged in with the correct WinUSB drivers not being detected in Ryu but being natively found in Dolphin. As it stands right now Ryujinx does not support GameCube controllers via the official adapter natively, SDL2 as a standalone library may but Ryujinx does not. 

The other issue is the addition of a GCC input option but I feel like the native implementation is of greater importance to this feature request.

https://user-images.githubusercontent.com/44103205/123871765-0627eb80-d92c-11eb-8c5d-888389e07d5f.mp4

**Comment by MutantAura** [comment]
Looked into this a bit further and potentially this info could help anyone who eventually takes an interest in this: https://github.com/libsdl-org/SDL/issues/4199

Seems like SDL does technically support this natively from this commit: https://github.com/libsdl-org/SDL/commit/c528615626d9f7789a5681a946cb3d5bd5d68c2c

**Comment by paerbac** [comment]
any updates? I really wanna play hdr remix with my gcc

**Comment by avahe-kellenberger** [comment]
Maybe off-topic, but why does Ryujinx not use system libraries? Or, is there a way to compile it to do so?
I have SDL2 compiled to support the Gamecube Controller, but cannot map any buttons under a `Pro Controller` in the input settings.

**Comment by edisionnano** [comment]
> Maybe off-topic, but why does Ryujinx not use system libraries? Or, is there a way to compile it to do so? I have SDL2 compiled to support the Gamecube Controller, but cannot map any buttons under a `Pro Controller` in the input settings.

Delete the provided libsdl2.so, this will make Ryujinx use the system one

**Comment by avahe-kellenberger** [comment]
Thanks, that worked.

For some info if anyone else is looking to do this:
- I found `libSDL2.so` in `/opt/ryujinx`
- The GCC shows up as a joycon
- Build sdl with `./configure --enable-hidapi-libusb && make -j install` for GCC support

**Comment by edisionnano** [comment]
> Thanks, that worked.
> 
> For some info if anyone else is looking to do this:
> 
>     * I found `libSDL2.so` in `/opt/ryujinx`
> 
>     * The GCC shows up as a joycon
> 
>     * Build sdl with `./configure --enable-hidapi-libusb && make -j install` for GCC support

Don't get Ryujinx from the AUR smh

**Comment by avahe-kellenberger** [comment]
Why?

**Comment by edisionnano** [comment]
> Why?

It isn't official

**Comment by paerbac** [comment]
> > Why?
> 
> It isn't official

Ryujinx has an MIT license, he can install it however he wants. The only "official" way is installing a tar.gz which is annoying bc it doesn't usually integrate with graphical package managers like pamac or bauh. Flatpak and AUR are just better in that regard.
All that being said... Still waiting for actual gcc input support, showing up as a joycon will still mean that the controller acts like a joycon to some degree, especially when it comes to analogue triggers and deadzones. A solution like this will not be adequate for a serious smash player that just wants to be able to play HDR remix online.

**Comment by EmulationFanatic** [comment]
> > > Why?
> > 
> > 
> > It isn't official
> 
> Ryujinx has an MIT license, he can install it however he wants. The only "official" way is installing a tar.gz which is annoying bc it doesn't usually integrate with graphical package managers like pamac or bauh. Flatpak and AUR are just better in that regard.
> 
> All that being said... Still waiting for actual gcc input support, showing up as a joycon will still mean that the controller acts like a joycon to some degree, especially when it comes to analogue triggers and deadzones. A solution like this will not be adequate for a serious smash player that just wants to be able to play HDR remix online.

The licensing has nothing to do with whether it's an official release from the Ryujinx dev team or not.

**Comment by avahe-kellenberger** [comment]
It's just an install script you can read, download, and run. It's fine, doesn't need to be official.

**Comment by edisionnano** [comment]
> > > Why?
> > 
> > 
> > It isn't official
> 
> Ryujinx has an MIT license, he can install it however he wants. The only "official" way is installing a tar.gz which is annoying bc it doesn't usually integrate with graphical package managers like pamac or bauh. Flatpak and AUR are just better in that regard.
> 
> All that being said... Still waiting for actual gcc input support, showing up as a joycon will still mean that the controller acts like a joycon to some degree, especially when it comes to analogue triggers and deadzones. A solution like this will not be adequate for a serious smash player that just wants to be able to play HDR remix online.

It's not about the license, the only Ryujinx package on aur is Ryujinx-git which doesn't provide a binary but rather builds Ryujinx from source. The first issue with that is that most likely you will have an old version of Ryujinx if you decide to use this package unless you force build it every day since the maintainer doesn't seem very active (last commit was on 15/9, that's about 25 versions behind since AUR helpers only rebuild git packages when there's a change). Official Ryujinx binaries also have an autoupdater which I believe this version lacks.

**Comment by MutantAura** [comment]
Can we keep issue channels on topic and not a chatroom please 👍 I'll be deleting anything further.

**Comment by Scrooge200FES** [comment]
Bumping this; I have a Mayflash adapter, what can I do? I tried the ViGEm Bus Driver solution, but installing the program made my laptop freeze up. I'd like to have it as a direct feature rather than jumping through so many hoops.

**Comment by TailsGuy** [comment]
Bumping this once again, It's about to be April, it'd be cool for this to be a thing.
Also, asynchronous shader compilation **SERIOUSLY** has to be considered...
